src/connector/src/parser/json_parser.rs

```diff
@@ -84,51 +87,79 @@ impl JsonParser {
     #[allow(clippy::unused_async)]
     pub async fn parse_inner(
         &self,
-        payload: Vec<u8>,
+        mut payload: Vec<u8>,
         mut writer: SourceStreamChunkRowWriter<'_>,
     ) -> Result<WriteGuard> {
-        enum Op {
-            Insert,
-            Delete,
-        }
-
-        let (mut payload_mut, op) = if self.enable_upsert {
+        if self.enable_upsert {
             let msg: UpsertMessage<'_> = bincode::deserialize(&payload)
                 .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
-            if !msg.record.is_empty() {
-                (msg.record.to_vec(), Op::Insert)
+
+            let mut primary_key = msg.primary_key.to_vec();
+            let mut record = msg.record.to_vec();
+            let key_value = simd_json::to_borrowed_value(&mut primary_key)
+                .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+
+            let value_value = if msg.record.is_empty() {
+                None
             } else {
-                (msg.primary_key.to_vec(), Op::Delete)
-            }
-        } else {
-            (payload, Op::Insert)
-        };
-
-        let value: BorrowedValue<'_> = simd_json::to_borrowed_value(&mut payload_mut)
-            .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
-
-        if let BorrowedValue::Array(ref objects) = value && matches!(op, Op::Insert) {
-            at_least_one_ok(
-                objects
-                    .iter()
-                    .map(|obj| Self::parse_single_value(obj, &mut writer))
-                    .collect_vec(),
-            )
-        } else {
-            let fill_fn = |desc: &SourceColumnDesc| {
-                simd_json_parse_value(
-                    &SourceFormat::Json,
-                    &desc.data_type,
-                    json_object_smart_get_value(&value, desc.name.as_str().into()),
+                Some(
+                    simd_json::to_borrowed_value(&mut record)
+                        .map_err(|e| RwError::from(ProtocolError(e.to_string())))?,
                 )
-                .map_err(|e| {
-                    tracing::error!("failed to process value: {}", e);
-                    e.into()
-                })
             };
-            match op {
-                Op::Insert => writer.insert(fill_fn),
-                Op::Delete => writer.delete(fill_fn),
+            let accessor = UpsertAccess {
```

Comment:
Why not provide an interface to easily construct a `UpdateAccess` for different encoding formats?

Reply:
According to the design, the UpdateAccess should know nothing about the encoding format. It works with the other components as long as they have implemented the Access trait. 
I have hidden the internal structure of UpsertAccess and provides constructors, please refresh and see it.